### PR TITLE
bitbucket and heptapod

### PR DESCRIPTION
**Pull request by jspaaks**
Added bitbucket suppport. Refs #52 

Wait with fixing this PR until #285 is merged

```text
howfairis https://bitbucket.org/atlassian/python-bitbucket
Checking compliance with fair-software.eu...
url: https://bitbucket.org/atlassian/python-bitbucket
(1/5) repository
      ✓ has_open_repository
(2/5) license
      × has_license
(3/5) registry
      × has_ascl_badge
      × has_bintray_badge
      × has_conda_badge
      × has_cran_badge
      × has_crates_badge
      × has_maven_badge
      × has_npm_badge
      ✓ has_pypi_badge
      × has_rsd_badge
      × is_on_github_marketplace
(4/5) citation
      × has_citation_file
      × has_citationcff_file
      × has_codemeta_file
      × has_zenodo_badge
      × has_zenodo_metadata_file
(5/5) checklist
      × has_core_infrastructures_badge

Calculated compliance: ● ○ ● ○ ○

It seems you have not yet added the fair-software.eu badge to your README.rst. You can do so by pasting the following snippet:

.. image:: https://img.shields.io/badge/fair--software.eu-%E2%97%8F%20%20%E2%97%8B%20%20%E2%97%8F%20%20%E2%97%8B%20%20%E2%97%8B-orange
   :target: https://fair-software.eu
```

Also added Heptapod support (#159):

```
howfairis https://foss.heptapod.net/fluiddyn/fluidlab
Checking compliance with fair-software.eu...
url: https://foss.heptapod.net/fluiddyn/fluidlab
(1/5) repository
      ✓ has_open_repository
(2/5) license
      ✓ has_license
(3/5) registry
      × has_ascl_badge
      × has_bintray_badge
      × has_conda_badge
      × has_cran_badge
      × has_crates_badge
      × has_maven_badge
      × has_npm_badge
      ✓ has_pypi_badge
      × has_rsd_badge
      × is_on_github_marketplace
(4/5) citation
      × has_citation_file
      × has_citationcff_file
      × has_codemeta_file
      × has_zenodo_badge
      × has_zenodo_metadata_file
(5/5) checklist
      × has_core_infrastructures_badge

Calculated compliance: ● ● ● ○ ○

It seems you have not yet added the fair-software.eu badge to your README.rst. You can do so by pasting the following snippet:

.. image:: https://img.shields.io/badge/fair--software.eu-%E2%97%8F%20%20%E2%97%8F%20%20%E2%97%8F%20%20%E2%97%8B%20%20%E2%97%8B-orange
   :target: https://fair-software.eu

```

### CODE REVIEW

File: howfairis/mixins/license_mixin.py [hunk 1]
```diff
@@ -20,6 +20,10 @@ def has_license(self):
 
         r = False
 
+        if self.repo.platform == Platform.BITBUCKET:
+            # I haven't found a good way to determine the license on BitBucket yet
+            r = False
```
Comment (sverhoeven) [inline howfairis/mixins/license_mixin.py:25]: Spinned of https://github.com/fair-software/howfairis/issues/183 to see if LICENSE file is present

**Review by sverhoeven — CHANGES_REQUESTED** [review]
Running against bitbucket or heptapod URLs gives expected compliance.

```shell
howfairis http://bitbucket.org/jspaaks/badge-test
howfairis https://bitbucket.org/verhoes/amifair
howfairis https://bitbucket.org/bitbucketpipelines/bitbucket-pipes-toolkit
howfairis https://foss.heptapod.net/mercurial/mercurial-bitbucketize
```

When platform does not match we should raise an exception ourselves instead of returning None and having later code raise a weird error.

For heptapod tests created #184

File: howfairis/mixins/license_mixin.py [hunk 2]
```diff
@@ -51,5 +55,25 @@ def has_license(self):
...
+                return r
+
+            r = response is not None and \
+                response.text is not None and  \
+                BeautifulSoup(response.text, "html.parser") is not None and \
+                BeautifulSoup(response.text, "html.parser") \
+                .find("div", class_="project-buttons") is not None and \
+                BeautifulSoup(response.text, "html.parser") \
+                .find("div", class_="project-buttons")\
+                .find(string="No license. All rights reserved") is None
```
[the remark this answers is not in the document]
Comment (jspaaks) [reply howfairis/mixins/license_mixin.py:76]: Can do, but I'd prefer to let the development be guided by the feedback we will be getting from SonarCloud. That will help take out ambiguity about what is or isnt a good pattern.

[the remark this answers is not in the document]
Comment (sverhoeven) [reply howfairis/mixins/license_mixin.py:76]: Agree, let's use the duplicate code warnings from SonarCloud.

File: howfairis/vcs_platform.py [hunk 3]
```diff
@@ -5,5 +5,7 @@
 
 @unique
 class Platform(Enum):
+    BITBUCKET = auto()
     GITHUB = auto()
     GITLAB = auto()
+    HEPTAPOD = auto()
```
[the remark this answers is not in the document]
Comment (jspaaks) [reply howfairis/vcs_platform.py:11]: I don't think this is the correct place for writing down that information. BTW if you go to foss.heptapod.net, it says "GitLab" at the top and looks basically like GitLab, so perhaps that's enough indication?

File: tests/bitbucket/jspaaks/badge_test/mocker.py [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+@pytest.fixture
+def mocker():
+    with requests_mock.Mocker() as m:
+        default_branch_json = {
+            "mainbranch": {
+                "type": "branch",
+                "name": "master"
+            }
+        }
+        m.get("http://bitbucket.org/jspaaks/badge-test")
```
[the remark this answers is not in the document]
Comment (jspaaks) [reply tests/bitbucket/jspaaks/badge_test/mocker.py:14]: Good catch on that https. I noticed this behavior as well, https://github.com/fair-software/howfairis/pull/171#issue-569684431, I think I know how to solve this. The problem is that we're making requests outside the mocker context, hence howfairis just goes online and does its thing. I will continue this work in the other PR, not here.

File: howfairis/vcs_platform.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (sverhoeven) [reply howfairis/vcs_platform.py:11]: OK, created https://github.com/fair-software/howfairis/issues/193 to have heptapod mentioned in the docs.

[the remark this answers is not in the document]
Comment (jspaaks) [reply howfairis/vcs_platform.py]: Added this to `cli`'s information string and in the README.rst

**Review by sverhoeven — APPROVED** [review]
All my suggestions where applied or delayed to other issues.

**Comment by jspaaks** [comment]
During the planning we put this PR out of scope for 0.13.0, where do we stand on that now?

**Comment by sverhoeven** [comment]
We will keep it out of scope for 0.13.0

**Comment by fdiblen** [comment]
Can we close this?

**Comment by jspaaks** [comment]
Closing is fine with me, it's now probably more work to sync this than to start anew.

**Pull request closed without merging**